vm/proxyapp/proxyappclient.go

```diff
@@ -0,0 +1,472 @@
+		}
+
+		pool.proxy.terminate()
+		<-pool.proxy.onTerminated
+
+		return fmt.Errorf("failed to construct pool: %w", err)
+	}
+
+	pool.count = count
+	return nil
+}
+
+type Config struct {
+	Command        string          `json:"cmd"`
+	ProxyAppConfig json.RawMessage `json:"config"`
+}
+
+func parseConfig(conf []byte) (*Config, error) {
+	vmCfg := new(Config)
+	if err := config.LoadData(conf, vmCfg); err != nil {
+		return nil, fmt.Errorf("failed to parseConfig(): %w", err)
+	}
+	return vmCfg, nil
+}
+
+// Package configuration VARs are mostly needed for tests.
+type proxyAppParams struct {
+	CommandRunner   func(context.Context, string, ...string) subProcessCmd
+	MakePipe        func() (io.ReadCloser, io.WriteCloser, error)
+	InitRetryDelay  time.Duration
+	MaxInitAttempts int64
+}
+
+func ctor(params *proxyAppParams, env *vmimpl.Env) (vmimpl.Pool, error) {
+	subConfig, err := parseConfig(env.Config)
+	if err != nil {
+		return nil, fmt.Errorf("config parse error: %w", err)
+	}
+
+	pool := &pool{
+		env: env,
+	}
+
+	initResult := make(chan error, 1)
+	pool.mu.Lock()
+
+	minWait := time.After(0)
+	go func() {
+		for attempt := int64(0); attempt < params.MaxInitAttempts; attempt++ {
+			<-minWait
+			minWait = time.After(params.InitRetryDelay)
+
+			err = initPool(params, pool, subConfig.Command, subConfig.ProxyAppConfig, env)
+			select {
+			case initResult <- err:
+			default:
+			}
+			if err != nil {
+				continue
+			}
+
+			pool.mu.Unlock()
+			<-pool.proxy.onTerminated
+			pool.mu.Lock()
+		}
+
+		pool.mu.Unlock()
+		select {
+		case initResult <- fmt.Errorf("max attempts(%v) error", params.MaxInitAttempts):
+		default:
+		}
+	}()
+
+	if err := <-initResult; err != nil {
+		return nil, fmt.Errorf("can't initialize pool: %w", err)
+	}
+	return pool, nil
+}
+
+func (pool *pool) Count() int {
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	return pool.count
+}
+
+func (pool *pool) Create(workdir string, index int) (vmimpl.Instance, error) {
+	pool.mu.Lock()
+	proxy := pool.proxy
+	pool.mu.Unlock()
+
+	return proxy.CreateInstance(workdir, index)
+}
+
+// Close is not used now. Its support require wide code changes.
+// TODO: support the pool cleanup on syz-manager level.
+func (pool *pool) Close() error {
+	// TODO: pool.proxy.Call("Close")
+	pool.proxy.terminate()
+	return nil
+}
+
+type ProxyApp struct {
+	*rpc.Client
+	terminate    context.CancelFunc
+	onTerminated chan bool
+	subProcess   subProcessCmd // is needed for mocking
+	params       *proxyAppParams
+}
+
+type stdInOutCloser struct {
+	io.ReadCloser
+	io.WriteCloser
+}
+
+func (s stdInOutCloser) Close() error {
+	readErr := s.ReadCloser.Close()
+	writeErr := s.WriteCloser.Close()
+	if readErr != nil || writeErr != nil {
+		return fmt.Errorf("stdInOutCloser.Close() error on read: %v, write: %v", readErr, writeErr)
+	}
+	return nil
+}
+
+func runProxyApp(params *proxyAppParams, cmd string) (*ProxyApp, error) {
+	ctx, cancelContext := context.WithCancel(context.TODO())
```

Comment:
Use context.Background, TODO is something that needs to be fixed.

Reply:
done